pipeline.cpp

```diff
@@ -78,9 +79,34 @@ void DrawLine(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* cons
  * Renders a triangle to the target buffer. Essential 
  * building block for most of drawing.
  ************************************************************/
-void DrawTriangle(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* const attrs, Attributes* const uniforms, FragmentShader* const frag)
+void DrawTriangle(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* const attrs, Attributes* const uniforms = NULL, FragmentShader* const frag = NULL)
 {
-    // Your code goes here
+    // Get the max and min boundaries of the triangle
+    int maxX = fmax(triangle[0].x, fmax(triangle[1].x, triangle[2].x));
+    int minX = fmin(triangle[0].x, fmin(triangle[1].x, triangle[2].x));
+    int maxY = fmax(triangle[0].y, fmax(triangle[1].y, triangle[2].y));
+    int minY = fmin(triangle[0].y, fmin(triangle[1].y, triangle[2].y));
+
+    // Vertexes for the cross product computations
+    Vertex vs1 = {triangle[1].x - triangle[0].x, triangle[1].y - triangle[0].y, 1, 1};
+    Vertex vs2 = {triangle[2].x - triangle[0].x, triangle[2].y - triangle[0].y, 1, 1};
+
+    for (int y = minY; y <= maxY; y++)
+    {
+        for (int x = minX; x <= maxX; x++)
+        {
+            // Use cross product to check if the pixel coordinates are in the bounds of the triangle.
+            Vertex q = {x - triangle[0].x, y - triangle[0].y, 1, 1};
+
+            float s = ((q.x*vs2.y)-(vs2.x*q.y)) / ((vs1.x*vs2.y)-(vs2.x*vs1.y));
+            float t = ((vs1.x*q.y)-(q.x*vs1.y)) / ((vs1.x*vs2.y)-(vs2.x*vs1.y));
+
+            if ((s >= 0) && (t >= 0) && (s + t <= 1))
+            {
+                target[y][x] = attrs[0].color;
```

Comment:
Looks good, good use of the reading example and changing it to work in this environment